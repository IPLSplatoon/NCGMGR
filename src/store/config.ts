import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const configStore = createStore<Configuration>({
    state: {
        installPath: ''
    },
    mutations: {
        setInstallPath: (state, installPath: string) => {
            state.installPath = installPath
        },
        setState: (state, config: Configuration) => {
            Object.assign(state, config)
        }
    },
    actions: {
        load: store => {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            store.commit('setState', {
                ...store.state,
                ...config
            })
        },
        persist: store => {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(store.state))
        }
    }
})

export const configStoreKey: InjectionKey<Store<Configuration>> = Symbol()

export function useConfigStore (): Store<Configuration> {
    return useStore(configStoreKey)
}
