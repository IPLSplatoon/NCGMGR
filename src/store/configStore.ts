import { defineStore } from 'pinia'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const useConfigStore = defineStore('config', {
    state: () => ({
        installPath: ''
    } as Configuration),
    actions: {
        load () {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            this.$patch(config)
        },
        persist () {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(this.$state))
        }
    }
})
