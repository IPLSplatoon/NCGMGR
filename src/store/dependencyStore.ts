import { defineStore } from 'pinia'
import { isBlank } from '@iplsplatoon/vue-components'
import { getNodejsVersion } from '@/service/dependencyService'

interface DependencyStore {
    isLoading: boolean
    nodejs: {
        version: string | null
    }
}

export const useDependencyStore = defineStore('dependency', {
    state: () => ({
        isLoading: false,
        nodejs: {
            version: null
        }
    } as DependencyStore),
    getters: {
        hasNodejs: state => !isBlank(state.nodejs.version)
    },
    actions: {
        async checkVersions () {
            this.isLoading = true
            try {
                this.nodejs.version = await getNodejsVersion()
            } finally {
                this.isLoading = false
            }
        }
    }
})
