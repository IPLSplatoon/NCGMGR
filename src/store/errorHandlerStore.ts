import { App } from 'vue'
import { defineStore } from 'pinia'
import { generateId } from '@/util/generateId'

interface ErrorHandlerStore {
    recentErrors: Record<string, unknown>
}

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    } as ErrorHandlerStore),
    actions: {
        removeRecentError ({ key }: { key: string }): void {
            delete this.recentErrors[key]
        },
        handleError ({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err)

            const id = generateId()
            this.recentErrors[id] = err
        }
    }
})

export function setUpErrorHandler (app: App<unknown>): void {
    const store = useErrorHandlerStore()
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info })
    }
}
