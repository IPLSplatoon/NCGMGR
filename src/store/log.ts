import { LogEvent } from '@/types/log'
import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'
import { Event, listen, UnlistenFn } from '@tauri-apps/api/event'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    completed: Record<string, boolean>
}

const unlistenFns: Record<string, UnlistenFn> = {}

export const logStore = createStore<LogStore>({
    state: {
        lines: {},
        completed: {}
    },
    mutations: {
        insertLine (state, { line, key }: { line: LogEvent, key: string }) {
            if (!state.lines[key]) {
                state.lines[key] = []
            }
            state.lines[key].push(line)
        },
        reset (state, key: string) {
            state.lines[key] = []
            state.completed[key] = false
        },
        setCompleted (state, { key, completed }: { key: string, completed: boolean }) {
            state.completed[key] = completed
        }
    },
    actions: {
        listen (store, key: string) {
            if (unlistenFns[key]) {
                return
            }
            return listen(`log:${key}`, (event: Event<LogEvent>) => {
                store.commit('insertLine', { line: event.payload, key })
            }).then(unlisten => {
                unlistenFns[key] = unlisten
            })
        },
        unlisten (store, key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key]()
            }
        },
        logPromiseResult (store, { promise, key, noLogOnSuccess }: { promise: Promise<unknown>, key: string, noLogOnSuccess: boolean }) {
            promise.then(() => {
                if (!noLogOnSuccess) {
                    store.commit('insertLine', {
                        line: {
                            message: 'Success!',
                            type: 'success'
                        },
                        key
                    })
                }
            }).catch(e => {
                store.commit('insertLine', {
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
            }).finally(() => {
                store.commit('setCompleted', { key, completed: true })
            })
        }
    }
})

export const logStoreKey: InjectionKey<Store<LogStore>> = Symbol()

export function useLogStore (): Store<LogStore> {
    return useStore(logStoreKey)
}
