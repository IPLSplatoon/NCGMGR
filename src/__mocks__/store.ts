import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        }
    })
}
