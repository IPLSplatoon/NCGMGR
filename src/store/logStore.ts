import { ActionState, LogEvent, ProgressEvent } from '@/types/log'
import { listen, UnlistenFn } from '@tauri-apps/api/event'
import { defineStore } from 'pinia'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    progressEntries: Record<string, ProgressEvent>
    actionStates: Record<string, ActionState>
}

const unlistenFns: Record<string, UnlistenFn[]> = {}

export const useLogStore = defineStore('log', {
    state: () => ({
        lines: {},
        progressEntries: {},
        actionStates: {}
    } as LogStore),
    actions: {
        insertLine ({ line, key }: { line: LogEvent, key: string }) {
            if (!this.lines[key]) {
                this.lines[key] = []
            }
            this.lines[key].push(line)
        },
        setProgress (key: string, event: ProgressEvent) {
            const existingEntry = this.progressEntries[key]
            if (existingEntry == null) {
                this.progressEntries[key] = event
                this.logProgressEvent(key, event)
            } else {
                existingEntry.message = event.message
                if (event.maxStep != null) {
                    existingEntry.maxStep = event.maxStep
                }
                if (event.step != null) {
                    existingEntry.step = event.step

                    if (event.step === event.maxStep) {
                        this.setActionState({ key, state: ActionState.COMPLETED_SUCCESS })
                    }
                }

                this.logProgressEvent(key, {
                    message: event.message,
                    step: event.step,
                    maxStep: existingEntry.maxStep
                })
            }
        },
        logProgressEvent(key: string, event: ProgressEvent) {
            if (event.step != null) {
                this.insertLine({
                    line: {
                        message: `[${event.step}/${event.maxStep}] ${event.message}`,
                        type: 'info'
                    },
                    key
                })
            } else {
                this.insertLine({
                    line: {
                        message: event.message,
                        type: 'info'
                    },
                    key
                })
            }
        },
        reset (key: string) {
            delete this.progressEntries[key]
            this.lines[key] = []
            this.actionStates[key] = ActionState.INCOMPLETE
        },
        setActionState ({ key, state }: { key: string, state: ActionState }) {
            this.actionStates[key] = state
        },
        listen (key: string, listenForProgress = false) {
            if (unlistenFns[key]) {
                return
            }

            const listenPromises: Array<Promise<UnlistenFn>> = []

            listenPromises.push(listen<LogEvent>(`log:${key}`, event => {
                this.insertLine({ line: event.payload, key })
            }))

            if (listenForProgress) {
                listenPromises.push(listen<ProgressEvent>(`progress:${key}`, event => {
                    this.setProgress(key, event.payload)
                }))
            }

            return Promise.all(listenPromises).then(result => {
                unlistenFns[key] = result
            })
        },
        unlisten (key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key].forEach(fn => fn())
                delete unlistenFns[key]
            }
        },
        logPromiseResult ({ promise, key }: { promise: Promise<unknown>, key: string }) {
            promise.catch(e => {
                this.insertLine({
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
                this.setActionState({ key, state: ActionState.COMPLETED_ERROR })
            })
        }
    }
})
