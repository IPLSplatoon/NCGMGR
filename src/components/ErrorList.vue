<template>
    <ipl-space class="error-list">
        <div class="m-y-8 m-x-8 text-center">
            <div class="m-b-8">
                Errors that have occurred in this application will be listed here.<br>
                It is most useful to developers for diagnosing issues.
            </div>
            <ipl-button
                :disabled="!errorsPresent"
                label="Clear"
                @click="errorHandlerStore.clear()"
            />
        </div>

        <ipl-expanding-space-group v-if="errorsPresent">
            <ipl-expanding-space
                v-for="(error, key) in errorHandlerStore.recentErrors"
                :key="key"
                :title="addDots(String(error.err))"
                class="error-list-item"
            >
                <ipl-data-row label="Info" :value="error.info" />
                <template v-if="error.component">
                    <ipl-data-row label="Source component" :value="error.component.$.type.name" />
                </template>
                <ipl-data-row label="Full error" :value="String(error.err)" />
            </ipl-expanding-space>
        </ipl-expanding-space-group>
        <div v-else class="m-l-8 m-y-8 text-center">No errors yet!</div>
    </ipl-space>
</template>

<script lang="ts">
import { defineComponent } from '@vue/runtime-core'
import { useErrorHandlerStore } from '@/store/errorHandlerStore'
import { IplButton, IplDataRow, IplExpandingSpace, IplExpandingSpaceGroup, IplSpace } from '@iplsplatoon/vue-components'
import { computed } from 'vue'
import { addDots } from '@/util/stringUtil'

export default defineComponent({
    name: 'ErrorList',

    components: { IplSpace, IplDataRow, IplButton, IplExpandingSpaceGroup, IplExpandingSpace },

    setup () {
        const errorHandlerStore = useErrorHandlerStore()

        return {
            errorHandlerStore,
            errorsPresent: computed(() => Object.keys(errorHandlerStore.recentErrors).length > 0),
            addDots
        }
    }
})
</script>

<style lang="scss" scoped>
.error-list {
    padding: 0;
}

.error-list-item {
    border-radius: 0;
    border: 0;
    border-top: 1px solid var(--space-border);
}
</style>
