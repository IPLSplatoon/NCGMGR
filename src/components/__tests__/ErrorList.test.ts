import { config, mount } from '@vue/test-utils'
import { createTestingPinia, TestingPinia } from '@pinia/testing'
import { useErrorHandlerStore } from '@/store/errorHandlerStore'
import ErrorList from '@/components/ErrorList.vue'

describe('ErrorList', () => {
    config.global.stubs = {
        IplButton: true,
        IplDataRow: true,
        FontAwesomeIcon: true
    }

    let pinia: TestingPinia

    beforeEach(() => {
        pinia = createTestingPinia()
        config.global.plugins = [pinia]
    })

    it('matches snapshot', () => {
        const errorHandlerStore = useErrorHandlerStore()
        errorHandlerStore.recentErrors = {
            err1: {
                err: new Error('Error!'),
                info: 'error information',
                component: {
                    // @ts-ignore
                    $: {
                        type: {
                            name: 'ComponentName'
                        }
                    }
                }
            },
            err2: {
                err: new Error('Error 2!'),
                info: 'info',
                component: null
            }
        }
        const wrapper = mount(ErrorList)

        expect(wrapper.html()).toMatchSnapshot()
    })
})
