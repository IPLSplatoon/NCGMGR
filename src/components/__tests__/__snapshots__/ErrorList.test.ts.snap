// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorList matches snapshot 1`] = `
<div>
  <div class="m-y-8 m-x-8 text-center">
    <div class="m-b-8"> Errors that have occurred in this application will be listed here.<br> It is most useful to developers for diagnosing issues. </div>
    <ipl-button-stub label="Clear" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href=""></ipl-button-stub>
  </div>
  <div class="ipl-expanding-space-group">
    <div class="ipl-expansion-panel__content error-list-item" data-v-67856a23="">
      <div class="ipl-expansion-panel__header" data-v-67856a23="">
        <div class="ipl-expansion-panel__header-background" data-v-67856a23=""></div>
        <div class="ipl-expansion-panel__title" data-v-67856a23="">Error: Error!</div>
        <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon" data-v-67856a23=""></font-awesome-icon-stub>
        <div class="header-extra" data-v-67856a23=""></div>
      </div>
      <div class="content" data-v-67856a23="" style="display: none;">
        <ipl-data-row-stub label="Info" value="error information"></ipl-data-row-stub>
        <ipl-data-row-stub label="Source component" value="ComponentName"></ipl-data-row-stub>
        <ipl-data-row-stub label="Full error" value="Error: Error!"></ipl-data-row-stub>
      </div>
    </div>
    <div class="ipl-expansion-panel__content error-list-item" data-v-67856a23="">
      <div class="ipl-expansion-panel__header" data-v-67856a23="">
        <div class="ipl-expansion-panel__header-background" data-v-67856a23=""></div>
        <div class="ipl-expansion-panel__title" data-v-67856a23="">Error: Error 2!</div>
        <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon" data-v-67856a23=""></font-awesome-icon-stub>
        <div class="header-extra" data-v-67856a23=""></div>
      </div>
      <div class="content" data-v-67856a23="" style="display: none;">
        <ipl-data-row-stub label="Info" value="info"></ipl-data-row-stub>
        <!--v-if-->
        <ipl-data-row-stub label="Full error" value="Error: Error 2!"></ipl-data-row-stub>
      </div>
    </div>
  </div>
</div>
`;
